snuba/query/parser/__init__.py

```diff
@@ -224,3 +240,186 @@ def apply_aliases(exp: Expression) -> Expression:
             return replace(exp, alias=exp.column_name)
 
     query.transform_expressions(apply_aliases)
+
+
+def _expand_aliases(query: Query) -> None:
+    """
+    Recursively expand all the references to aliases in the query. This
+    makes life easy to query processors and translators that only have to
+    take care of not introducing shadowing (easy to enforce at runtime),
+    otherwise aliasing is transparent to them.
+    """
+    # Pre-inline all nested aliases. This reduces the number of iterations
+    # we need to do on the query.
+    # Example:
+    # {"a": f(x), "x": g(k)} -> {"a": f(g(k)), "x": g(k)}
+    aliased_expressions = {
+        exp.alias: exp for exp in query.get_all_expressions() if exp.alias is not None
+    }
+    fully_resolved_aliases = {
+        alias: exp.accept(
+            AliasExpanderVisitor(aliased_expressions, [], expand_nested=True)
+        )
+        for alias, exp in aliased_expressions.items()
+    }
+
+    visitor = AliasExpanderVisitor(fully_resolved_aliases, [])
+    query.set_ast_selected_columns(
+        [e.accept(visitor) for e in (query.get_selected_columns_from_ast() or [])]
+    )
+    arrayjoin = query.get_arrayjoin_from_ast()
+    if arrayjoin is not None:
+        query.set_ast_arrayjoin(arrayjoin.accept(visitor))
+    condition = query.get_condition_from_ast()
+    if condition is not None:
+        query.set_ast_condition(condition.accept(visitor))
+    query.set_ast_groupby(
+        [e.accept(visitor) for e in (query.get_groupby_from_ast() or [])]
+    )
+    having = query.get_having_from_ast()
+    if having is not None:
+        query.set_ast_having(having.accept(visitor))
+    query.set_ast_orderby(
+        list(
+            map(
+                lambda clause: replace(
+                    clause, expression=clause.expression.accept(visitor)
+                ),
+                query.get_orderby_from_ast(),
+            )
+        )
+    )
+
+
+class AliasExpanderVisitor(ExpressionVisitor[Expression]):
```

Comment:
See tests/query/parser/test_resolver_visitor.py for a coverage of the corner cases